storage/src/qmdb/any/db.rs

```diff
@@ -0,0 +1,502 @@
+/// - [crate::qmdb::any::ordered::Fixed]
+/// - [crate::qmdb::any::ordered::Variable]
+/// - [crate::qmdb::any::unordered::Fixed]
+/// - [crate::qmdb::any::unordered::Variable]
+pub struct Db<
+    E: Storage + Clock + Metrics,
+    C: Contiguous,
+    I,
+    H: Hasher,
+    U,
+    S: State<DigestOf<H>> = Clean<DigestOf<H>>,
+> where
+    C::Item: Codec,
+{
+    /// A (pruned) log of all operations in order of their application. The index of each
+    /// operation in the log is called its _location_, which is a stable identifier.
+    ///
+    /// # Invariants
+    ///
+    /// - The log is never pruned beyond the inactivity floor.
+    /// - There is always at least one commit operation in the log.
+    pub(crate) log: AuthenticatedLog<E, C, H, S>,
+
+    /// A location before which all operations are "inactive" (that is, operations before this point
+    /// are over keys that have been updated by some operation at or after this point).
+    pub(crate) inactivity_floor_loc: Location,
+
+    /// The location of the last commit operation.
+    pub(crate) last_commit_loc: Location,
+
+    /// A snapshot of all currently active operations in the form of a map from each key to the
+    /// location in the log containing its most recent update.
+    ///
+    /// # Invariant
+    ///
+    /// - Only references `Operation::Update`s.
+    pub(crate) snapshot: I,
+
+    /// The number of _steps_ to raise the inactivity floor. Each step involves moving exactly one
+    /// active operation to tip.
+    pub(crate) steps: u64,
+
+    /// The number of active keys in the snapshot.
+    pub(crate) active_keys: usize,
+
+    /// Marker for the update type parameter.
+    pub(crate) _update: core::marker::PhantomData<U>,
+}
+
+impl<E, K, V, U, C, I, H, S> Db<E, C, I, H, U, S>
+where
+    E: Storage + Clock + Metrics,
+    K: Array,
+    V: ValueEncoding,
+    U: Update<K, V>,
+    C: Contiguous<Item = Operation<K, V, U>>,
+    I: UnorderedIndex<Value = Location>,
+    H: Hasher,
+    S: State<DigestOf<H>>,
+    Operation<K, V, U>: Codec,
+{
+    /// The number of operations that have been applied to this db, including those that have been
+    /// pruned and those that are not yet committed.
+    pub fn op_count(&self) -> Location {
+        self.log.size()
+    }
+
+    /// Return the inactivity floor location. This is the location before which all operations are
+    /// known to be inactive. Operations before this point can be safely pruned.
+    pub const fn inactivity_floor_loc(&self) -> Location {
+        self.inactivity_floor_loc
+    }
+
+    /// Whether the snapshot currently has no active keys.
+    pub const fn is_empty(&self) -> bool {
+        self.active_keys == 0
+    }
+
+    /// Returns the location of the oldest operation that remains retrievable.
+    pub fn oldest_retained_loc(&self) -> Option<Location> {
+        self.log.oldest_retained_loc()
```

Comment:
The return value need not be an option (there should always be one retained commit).
(I know this is my code, just noticed it now!)